controllers/toolchaincluster/toolchaincluster_controller.go

```diff
@@ -72,23 +76,99 @@ func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.
-		remoteClusterClientset: clientSet,
-		logger:                 reqLogger,
-	}
+
+	// execute healthcheck
+	healthcheckResult := r.getClusterHealthCondition(ctx, clientSet)
+
 	// update the status of the individual cluster.
-	if err := healthChecker.updateIndividualClusterStatus(ctx, toolchainCluster); err != nil {
+	if err := r.updateStatus(ctx, toolchainCluster, healthcheckResult...); err != nil {
 		reqLogger.Error(err, "unable to update cluster status of ToolchainCluster")
 		return reconcile.Result{}, err
 	}
-
 	return reconcile.Result{RequeueAfter: r.RequeAfter}, nil
 }
 
+func (r *Reconciler) updateStatus(ctx context.Context, toolchainCluster *toolchainv1alpha1.ToolchainCluster, currentconditions ...toolchainv1alpha1.Condition) error {
+	toolchainCluster.Status.Conditions = condition.AddOrUpdateStatusConditionsWithLastUpdatedTimestamp(toolchainCluster.Status.Conditions, currentconditions...)
+	if err := r.Client.Status().Update(ctx, toolchainCluster); err != nil {
+		return errors.Wrapf(err, "Failed to update the status of cluster %s", toolchainCluster.Name)
+	}
+	return nil
+}
+
+func (r *Reconciler) getClusterHealthCondition(ctx context.Context, remoteClusterClientset *kubeclientset.Clientset) []v1alpha1.Condition {
+
+	healthcheck, errhealth := r.getClusterHealth(ctx, remoteClusterClientset)
+	if errhealth != nil {
+		return []v1alpha1.Condition{clusterOfflineCondition()}
+	}
+	if !healthcheck {
+		return []v1alpha1.Condition{clusterNotReadyCondition(), clusterNotOfflineCondition()}
+	}
+	return []v1alpha1.Condition{clusterReadyCondition()}
+
+}
+
+func (r *Reconciler) getClusterHealth(ctx context.Context, remoteClusterClientset *kubeclientset.Clientset) (bool, error) {
+	if r.CheckHealth != nil {
+		return r.CheckHealth(ctx, remoteClusterClientset)
+	}
+	return getClusterHealthStatus(ctx, remoteClusterClientset)
+}
+
+func clusterReadyCondition() toolchainv1alpha1.Condition {
+	currentTime := metav1.Now()
+	return toolchainv1alpha1.Condition{
+		Type:               toolchainv1alpha1.ConditionReady,
+		Status:             corev1.ConditionTrue,
+		Reason:             toolchainv1alpha1.ToolchainClusterClusterReadyReason,
+		Message:            healthzOk,
+		LastUpdatedTime:    &currentTime,
+		LastTransitionTime: currentTime,
+	}
+}
+
+func clusterNotReadyCondition() toolchainv1alpha1.Condition {
+	currentTime := metav1.Now()
+	return toolchainv1alpha1.Condition{
+		Type:               toolchainv1alpha1.ConditionReady,
+		Status:             corev1.ConditionFalse,
+		Reason:             toolchainv1alpha1.ToolchainClusterClusterNotReadyReason,
+		Message:            healthzNotOk,
+		LastUpdatedTime:    &currentTime,
+		LastTransitionTime: currentTime,
+	}
+}
+
+func clusterOfflineCondition() toolchainv1alpha1.Condition {
+	currentTime := metav1.Now()
+	return toolchainv1alpha1.Condition{
+		Type:               toolchainv1alpha1.ToolchainClusterOffline,
+		Status:             corev1.ConditionTrue,
```

Comment:
As I mentioned above, it would be nice to drop this condition to make it more consistent with the other controllers/CRDs
```suggestion
		Type:               toolchainv1alpha1.ConditionReady,
		Status:             corev1.ConditionFalse,
```
it would be also much easier to manage the changes of the states.

Reply:
it is same as condition Not ready, so you mean to drop clusteroffline condition itself..?

Reply:
updated it